Update function toString

MSLaguana opened the pull request:
According to https://tc39.github.io/Function-prototype-toString-revision/
the result of toString on a function should be the source text which matches
the grammar describing the kind of function. In practice this essentially
means that toString should just be whatever the source code defining the function
is.

To support this, we need to slightly adjust the extent of a function in
the case of member list declarations to ensure that it captures the start
of the function, including any `async` or computed function name, etc.

Fixes #2914

MSLaguana commented:
One of the tests is giving me trouble with newlines in the output compared to the baseline; I'm interested to see what Jenkins CI finds here.

sethbrenith commented on lib/Runtime/Library/ScriptFunction.cpp:
```diff
@@ -495,7 +495,7 @@ using namespace Js;
             else
             {
-                cachedSourceString = FormatToString(builder.ToString());
+                cachedSourceString = builder.ToString();// FormatToString(builder.ToString());
```
remove comment?

pleath commented on lib/Parser/Parse.cpp:
```diff
@@ -11510,11 +11521,52 @@ ParseNodeProg * Parser::Parse(LPCUTF8 pszSrc, size_t offset, size_t length, char
+
             // Eat the computed name expression
             if (m_token.tk == tkLBrack && isMethod)
             {
```
Is this part of the change solving a new problem that's introduced by changing the extents of the function to include the computed name? It seems like the problems you're describing are not new, so were we just not coming in here?

pleath replied:
Interesting. I see why matching brackets seems like it should work. I just wonder if it's really bulletproof. Is there a place where we can record the endpoint of the computed name so we can either advance the scanner past it or do a debug-only check that the scanner is arriving at the same place? Maybe a field on the DeferredStub?

boingoing replied:
Seems like it should parse correctly again if we upfront parsed it already. Agree it would be cool to assert the expression.

pleath replied:
Actually, now that I think of it, I'm worried about this. You can't actually tokenize JS without knowing where expressions begin and end, because that's how we resolve the ambiguous meaning of the forward slash. I wonder if we may really need to use ParseExpr and do what's necessary a. to recognize that the computed name expression belongs to the enclosing context, b. to recall whether the enclosing context is async.

MSLaguana replied:
Is there enough ambiguity that it would impact finding the next `]`? I had assumed that the tokenizer would handle strings and comments and regex literals, but that otherwise I'd get a sequence of fairly raw tokens, and I'm not aware of any compound tokens that include `[` or `]` that would need to depend on surrounding context.

pleath replied:
Regex literals are your problem, since those are marked by /. They can’t be tokenized w/o understanding expressions.

MSLaguana replied:
Ah, so the tokenizer doesn't automatically switch into regex tokenizing mode, like it does for strings? I'll have a play around tomorrow and see if something like `{ [/]/.exec("]")]() {}}` causes issues.

MSLaguana replied:
Looks like you are right, the above causes a syntax error when defer parsed.

MSLaguana replied:
I've just added handling for the regexp case, as well as a debug mode sanity check. Note that the sanity check isn't perfect; without the regex handling you could fool it with something like `{ [/]() { return 1}/.toString()]() { return 2;}}` and the defer parse leg will latch onto the string inside the regexp, thinking it to be legitimate.

MSLaguana replied:
Turns out it won't be considered legitimate, since it expects that once it has parsed the function it should have consumed all of the source, and in the above case it hasn't, so it throws a syntax error. So, if things don't work as we expect that would lead to a bug (runtime syntax error when invoking a correct function) but not otherwise incorrect behavior.

pleath replied:
So the idea is to decide whether / indicates regex or division based on the last token seen? Does that always work? Can function expressions, lambdas, etc. create confusion? Both syntax coloring and parallel parsing wanted to disambiguate these cases and never managed it to anybody's satisfaction. I'd honestly feel more comfortable having the initial parse record the point where the name ends and just moving the scanner past it.

sethbrenith replied:
Great points Paul; this logic has gotten very complicated. I'm also a bit worried about the possible bad outcome: a runtime syntax error is certainly better than a security bug, but worse than a controlled crash (because crawler would presumably find the crash for us).

MSLaguana replied:
The check that I added here was taken from `FastScanFormalsAndBody` which is one of the few places that switches the parser into regexp mode.

pleath replied:
Ah. That explains why I was suspicious of it: because I probably wrote it.

boingoing replied:
This is throwing several wrenches in. Pretty much agree with Paul, we should really keep track of the computed name expression and skip it that way.

boingoing commented on test/es6/computedPropertyToString.js:
```diff
@@ -34,15 +35,27 @@ var tests = [
+           assert.areEqual("null",        quxObj[name5()].name, "name5() returns null as the name");
+           assert.areEqual("true",        quxObj[name6()].name, "name6() returns true as the name");
+           assert.areEqual("false",       quxObj[name7()].name, "name7() returns false as the name");
+           //assert.areEqual(sym,         quxObj[name8()].name, "name8() returns false as the name"); // Symbol doesn't play nicely
```
Does this throw because implicit symbol conversion to a string throws?

MSLaguana replied:
Yeah, I believe that's what happened. The original test case just left 8 out entirely, which confused me for a moment which is why I added it here with a note.

MSLaguana commented:
@dotnet-bot test Windows 10 ci_slow_x64_debug please

MSLaguana commented:
Ok, after a lot of back-and-forth with @boingoing I've got an update here which should solve things I hope!

I've added a new field tracking the "to string" start of a function, and changed the general semantics of the `cbMin` to be "this is where parsing starts", while `cbStringMin` is "this is where `toString`-ing starts". This is then plumbed through from the `ParseNodeFnc`s to the `ParseableFunctionInfo` which can then differentiate between `GetSource`, which operates the same way as before since it's used for things other than `toString`, and `GetToStringSource` which now uses this new offset.

The `toString` method then has to do a little bit of gymnastics to use this new value: To avoid bloating memory, I've only added one new 4 byte value (after shuffling some things around so it will fit), and so I can't keep track of the byte-start and the char-start, so I don't know up-front how many utf16 chars the source will be. However in the worst case there is 1 per utf8 byte, so we potentially overprovision the `toString`-string. This will generally only be by a little bit, unless there are lots of non-ascii characters in something like a computed name expression of a function.

MSLaguana commented:
One other change with this: The `ParseNodeFnc` nodes should always have their `cbMin` as where parsing should start, and so async and generator functions don't have a special case there, they just have to update the `cbStringMin` for their `toString` representation. Since we still need to know whether these functions are async or generator at defer-parse time, I had to add some flags to pass that along.

If we can find another ~4 bytes somewhere, then we can also update class constructors to store the `toString`-end, and change their `cbLim` to reflect what needs to be parsed for just the constructor, and use the `cbStringMin` and a new `cbStringLim` for the purposes of `toString`, so class constructors could be deferred then.

MSLaguana commented:
@pleath please take another look at this and let me know what you think.

pleath commented on lib/Runtime/Library/ScriptFunction.cpp:
```diff
@@ -325,117 +325,6 @@ using namespace Js;
         return this->GetFunctionProxy()->EnsureDeserialized()->GetCachedSourceString();
     }
 
```
Love it that we can live without this now.

MSLaguana commented on lib/Runtime/Library/ScriptFunction.cpp:
```diff
@@ -474,12 +363,22 @@ using namespace Js;
+            size_t decodedCount = utf8::DecodeUnitsInto(builder.DangerousGetWritableBuffer(), pbStart, pbStart + cbLength + cbPreludeLength, options);
 
-            if (decodedCount != cch)
+            if (decodedCount < cch)
```
I need to fix this up a little actually; when we overestimate we still need to go and update the length. May as well copy into a new properly allocated string in that case.

MSLaguana replied:
Ok, that should be sorted now, although I'd appreciate a second set of eyes on this.

pleath replied:
Just out of curiosity, what’s the case in which we’re wrong about the length?

MSLaguana replied:
Because I'm only tracking one number, I only know the byte-offset of the start of the "To-string" part of the source, I don't know the character-offset. This means that if there are characters outside of 7 bit ascii, then we don't have "1-byte = 1 character" any more, so we will allocate too many characters' worth of space.

For example in the test I've added I use the ß character which is 2 bytes in utf8. If it appears somewhere like `async /* ß */ foo() {}` as part of a member function, the bounds that we track fully are for parsing, and so are for `foo() {}` or just `() {}`, I'd have to go double check again to say which exactly. We do have a byte-offset to the start of the `async` and so we can fully recompute the string, but there are ~15 byte difference between the byte indexes, and only ~14 characters in that region (again, because the ß is 2 bytes for one character).

MSLaguana commented on test/Basics/VerifySkipNestedDeferred.baseline:
```diff
@@ -12,6 +12,8 @@ SkipNestedDeferred:	Pushing a reference to 'x'
 Skipping nested deferred function 11. a1: 1438...1546
 SkipNestedDeferred:	Pushing a reference to 'x'
+Skipping nested deferred function 12. a2: 1469...1517
+SkipNestedDeferred:	Pushing a reference to 'x'
```
@boingoing can you take a look at this? I'm not sure what exactly this test is intending to do, and I wasn't expecting to make this kind of change, but it looks plausible to me...

boingoing replied:
This test is just verifying that when we have the parser state cache we can use it to skip scanning nested functions. What's going on here is we are skipping scan for a function which we weren't before. I guess this is because of the new flag you added for async functions in defer parse which is making us go down a different path in the parser. I didn't expect that but the test behavior here is not wrong so it should be fine as long as other tests are also passing.

MSLaguana commented:
Thanks for the reviews everyone! Once the Full run that I've kicked off completes successfully, I'll get this checked in.
